README.md
licence-picker
==============

Defines a licence taxonomy, so editors can set licences on a post-by-post basis. Prints the licencing meta tag. Use it for Creative Commons, GNU GPL etc.

* *Contributors:* Júlio Reis
* *Tags:* licensing, GPL, Creative Commons, metadata
* *Requires at least:* 3.0.0
* *Tested up to:* 3.8.1
* *Stable tag:* 0.0.1
* *License:* GPLv3
* *License URI:* [http://www.gnu.org/licenses/gpl-3.0.html](http://www.gnu.org/licenses/gpl-3.0.html) or see the included `LICENSE.txt` and `LICENSE.md`

Description
-----------

![Licence picker taxonomy](assets/screenshot-picker.png)

Defines a `licence` custom taxonomy. The licence info should be registered in this way:

* short name as `name` attribute (what you input when you create a licence from Add New Post / Edit Post)
* long name, followed by a space, followed by the licence’s URL as `description` attribute (accessible in the _Posts > Licences_ admin menu)

Then just pick the licences you need from the taxonomy box:

### Tested with

* [Polylang](http://polylang.wordpress.com/) – just turn on translations for the `licence` taxonomy under _Settings > Languages > Settings > Custom taxonomies_

### Suggested licences

These are some licences to get you started.

* `name` : `description`
* CC0 : Public Domain Dedication http://creativecommons.org/publicdomain/zero/1.0/
* CC BY 4.0 : Creative Commons Attribution 4.0 International http://creativecommons.org/licenses/by/4.0/
* CC BY-SA 4.0 : Creative Commons Attribution-ShareAlike 4.0 International http://creativecommons.org/licenses/by-sa/4.0/
* CC BY-ND 4.0 : Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International http://creativecommons.org/licenses/by-nd/4.0/
* CC BY-NC 4.0 : Creative Commons Attribution-NonCommercial 4.0 International http://creativecommons.org/licenses/by-nc/4.0/
* CC BY-NC-SA 4.0 : Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International http://creativecommons.org/licenses/by-nc-sa/4.0/
* CC BY-NC-ND 4.0 : Creative Commons Attribution-NonCommercial-NoDerivatives 4.0 International http://creativecommons.org/licenses/by-nc-nd/4.0/
* GNU GPL : GNU General Public License v3 http://www.gnu.org/licenses/gpl-3.0.html
* GNU Lesser GPL : GNU Lesser General Public License http://www.gnu.org/licenses/lgpl.html
* GNU Affero GPL : GNU Affero General Public License http://www.gnu.org/licenses/agpl-3.0.html
* GNU FDL : GNU Free Documentation License http://www.gnu.org/licenses/fdl.html
* PD : Public Domain Mark http://creativecommons.org/publicdomain/mark/1.0/

If you add licences programmatically with `wp_insert_term()`, remember to add them _only once_ or you’ll get duplicate terms.

Installation
------------

1. Upload the `licence-picker` folder to the `/wp-content/plugins/` directory
1. Activate the plugin through the ‘Plugins’ menu in WordPress
1. Go to _Posts > Licence _ to configure it if necessary

### Updates

This plugin works with `github-updater`: turn it on to get automatic update notices when a new version is published on GitHub.

### Translations

This plugin’s translations are managed via Transifex. To help out, please go to [https://www.transifex.com/projects/p/licence-picker/](https://www.transifex.com/projects/p/licence-picker/). Thank you!

Options
-------

* `LPBase::generate_licence_link` returns the HTML for a list of licences defined for the current post. If for single post pages, it defines the `rel="license"` attribute which associates the page with its licence.

Change log
----------

See [CHANGES.md](CHANGES.md)


CHANGES.md
Change log for Licence Picker
-----------------------------

#### 0.0.1

*2014-02-22*

* Added: Integration with [Transifex](https://www.transifex.com/projects/p/licence-picker/) for community translation
* Added: Incomplete translations: fr_FR, pt_BR
* Fixed: Path to pt_PT translation
* Fixed: Do not fire admin during an ajax request

#### 0.0.0

*2014-02-08*

* Initial release
* l10n: Portuguese (Portugal)


admin/class-lp-admin.php
<?php
/**
 * class-lp-admin.php
 * Admin class for plugin licence-picker
 */

// If this file is called directly, abort
if ( !defined('DB_NAME') ) {
	header('HTTP/1.0 403 Forbidden');
	die;
}

/**
 * Class that holds most of the admin functionality.
 */
class LP_Admin extends LP_Base {

	public function __construct() {
		add_action( 'admin_init', array( $this, 'requires_wordpress_version') );
		register_activation_hook( __FILE__, array( $this, 'activate' ) );
		register_deactivation_hook( __FILE__, array( $this, 'deactivate' ) );
	}
	
	/**
	 * To be run when the plugin is activated
	 * @return void
	 */
	static function activate() {
	}

	/**
	 * To be run when the plugin is deactivated
	 * @return void
	 */
	static function deactivate() {
	}

	/**
	 * Checks if the current WP install is newer than $wp_version
	 * Must test every time it's run, not just upon activation,
	 * because we don't know if the plugin code has been upgraded
	 * @return void
	 */
	public function requires_wordpress_version() {
		global $wp_version;
		$plugin = LP_BASENAME;
		$plugin_data = get_plugin_data( LP_MAINFILE, false );

		if ( version_compare($wp_version, LP_MIN_WP_VERSION, "<" ) ) {
			if( is_plugin_active( $plugin ) ) {
				// deactivate plugin, print error message
				deactivate_plugins( $plugin );
				// Translators: 1 is plugin name, 2 is version number
				$msg_title = sprintf( __( '%1$s %2$s not activated', 'licence-picker' ), $plugin_data['Name'], $plugin_data['Version'] );
				// Translators: 1 is current WordPress version, 2 is required version
				$msg_para = sprintf( __( 'You are running WordPress version %1$s. This plugin requires version %2$s or higher, and has been deactivated! Please upgrade WordPress and try again.', 'licence-picker' ), $wp_version, LP_MIN_WP_VERSION );
				$msg_back = __( 'Back to WordPress admin', 'licence-picker' );
				wp_die(  sprintf( '<h1>%s</h1><p>%s</p><p><a href="%s">%s</a></p>' , $msg_title, $msg_para, admin_url(), $msg_back ) );
			}
		}
	}
}

// Globalize the var first as it's needed globally
global $lp_admin;
$lp_admin = new LP_Admin();

// Register activation/deactivation
register_activation_hook( LP_MAINFILE, array( 'LP_Admin', 'activate' ) );
register_deactivation_hook( LP_MAINFILE, array( 'LP_Admin', 'deactivate' ) );


public/class-lp-frontend.php
<?php
/**
 * @package Frontend
 */

if ( !defined( 'LP_VERSION' ) ) {
	header( 'HTTP/1.0 403 Forbidden' );
	die;
}

class LP_Frontend extends LP_Base {

	public function __construct() {
	}

	/* Prints all of the licences for the current post
	 * @param str $sep  The separator between each licence
	 * @param str $glue The separator between the label and the list of licences
	 * @return Nothing
	 */
	static function the_licences( $sep, $glue ) {
		// display the licences, if any
		$licences = get_the_terms( get_the_id(), 'licence' );
		if( $licences ) {
			// collate all the terms
			$number = 0;
			$imploded_licences = '';
			foreach( $licences as $licence ) {
				$imploded_licences .= ( $number ? $glue : '' )
					. LP_Frontend::generate_licence_link( $licence );
				$number++;
			}
			// get label and print
			// Translators: The term 'licencing' in English does not require plural; do as works best for your language
			$label = _n( 'Licencing', 'Licencing', $number, 'licence-picker' );
			echo( "<li><label class=\"meta-label\">$label$sep</label>$imploded_licences</li>\n" );
		}
	}
	
	/* Returns the HTML for a link to the specified licence
	 * @param obj $licence The licence data
	 * @return str         The HTML
	 */
	static function generate_licence_link( $licence ) {
		// obtain long name and URL from description
		// long name is all text until the last space;
		// URL is what comes after the last space
		$space_pos = mb_strrpos( $licence->description, ' ' );
		if( $space_pos === false ) {
			$long_name = '';
			$url = $licence->description;
		}
		else {
			$long_name = mb_substr( $licence->description, 0, $space_pos );
			$url = mb_substr( $licence->description, $space_pos + 1 );
		}
		// compose the link if we have a URL, with title if we have a long name
		// if the licence is about this page, add a rel="license"
		$title = $long_name ? " a title=\"$long_name\"" : '';
		$rel = is_single() ? ' rel="license"' : '';
		$a_open = $url ? "<a$rel$title href=\"$url\">" : '';
		$a_close = $url ? '</a>' : '';
		return $a_open . $licence->name . $a_close;
	}
	
}

global $lp_front;
$lp_front = new LP_Frontend;

licence-picker.php
<?php 
/**
 * @package   Licence Picker
 * @author    Júlio Reis <webmaster@arocha.org>
 * @license   GPL-3.0
 * @link      http://arocha.org
 * @copyright 2014 A Rocha International
 *
 * @wordpress-plugin
 * Plugin Name:       Licence Picker
 * Plugin URI:        @TODO
 * Description:       Defines a licence taxonomy, so editors can set licences on a post-by-post basis. Prints the licencing meta tag. Use it for Creative Commons, GNU etc.
 * Version:           0.0.1
 * Author:            Júlio Reis / A Rocha International
 * Author URI:        http://www.tintazul.com.pt/julio.reis/
 * Text Domain:       licence-picker
 * License:           GPL-3.0
 * License URI:       http://www.gnu.org/licenses/gpl-3.0.txt
 * Domain Path:       /languages
 * GitHub Plugin URI: https://github.com/Tintazul/licence-picker
 * Requires at least: 3.0.0
 */

// If this file is called directly, abort
if ( !defined('DB_NAME') ) {
	header('HTTP/1.0 403 Forbidden');
	die;
}

define( 'LP_MAINFILE', __FILE__ );                    // absolute path to this file
define( 'LP_BASENAME', plugin_basename( __FILE__ ) ); // relative path to this file
define( 'LP_PATH', plugin_dir_path( __FILE__ ) );     // absolute path to plugin dir
define( 'LP_URL', plugin_dir_url( __FILE__ ) );       // URL to plugin dir
define( 'LP_VERSION', '0.0.1' );                      // plugin version
define( 'LP_MIN_WP_VERSION', '3.0.0');                // required minimum WP version

load_plugin_textdomain( 'licence-picker', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );

/**
 * Initialize LP admin functionality
 * @return void
 */
function lp_admin_init() {
	require LP_PATH . 'admin/class-lp-base.php';
	require LP_PATH . 'admin/class-lp-admin.php';
}

/**
 * Initialize LP frontend functionality
 * @return void
 */
function lp_frontend_init() {
	require LP_PATH . 'admin/class-lp-base.php';
	require LP_PATH . 'public/class-lp-frontend.php';
}

if ( is_admin() ) {
	if ( defined('DOING_AJAX') && DOING_AJAX ) {  // if we had ajax, we’d load it here
	}
	else add_action( 'plugins_loaded', 'lp_admin_init', 0 );
} else {	
	add_action( 'plugins_loaded', 'lp_frontend_init', 0 );
}
add_action( 'init', 'lp_create_taxonomy', 0 );

/* create the custom taxonomy
 * kudos to http://www.wpbeginner.com/wp-tutorials/create-custom-taxonomies-wordpress/
 * @return bool True if the taxonomy was created, false if not
 */
function lp_create_taxonomy() {
	// do nothing if taxonomy already exists
	if( taxonomy_exists( 'licence' ) )
		return false;
	// Add new taxonomy, make it hierarchical like categories
	// first do the translations part for GUI
	$labels = array(
		'name' => _x( 'Licence picker', 'taxonomy general name', 'licence-picker'),
		'singular_name' => _x( 'Licence picker', 'taxonomy singular name', 'licence-picker'),
		'search_items' =>  __( 'Search licences', 'licence-picker'),
		'all_items' => __( 'All licences', 'licence-picker'),
		'parent_item' => __( 'Parent licence', 'licence-picker'),
		'parent_item_colon' => __( 'Parent licence:', 'licence-picker'),
		'edit_item' => __( 'Edit licence', 'licence-picker'),
		'update_item' => __( 'Update licence', 'licence-picker'),
		'add_new_item' => __( 'Add New licence', 'licence-picker'),
		'new_item_name' => __( 'New licence', 'licence-picker'),
		'menu_name' => __( 'Licences', 'licence-picker'),
	);
	// Register the taxonomy
	register_taxonomy( 'licence', array('post'), array(
		'hierarchical' => true,
		'labels' => $labels,
		'show_ui' => true,
		'show_admin_column' => true,
		'query_var' => true,
		'rewrite' => array( 'slug' => 'licence' ),
	));
	// taxonomy was created
	return true;
}
